### PR TITLE
fix(unzip): Not all ZIP files have proper directory entries

### LINKED ISSUE #543: FileNotFoundException with certain ZIP files
### Description

Certain ZIP files cause an exception in `io.fabric8.launcher.base.Paths.unzip()`.

### What should happen?

No error ;)

### What happens instead?

Error ;)

### Do you have any idea why?

The current code assumes that each file entry in a ZIP will be preceded by a directory entry so it can create the directory before any files that need to be placed in it. It seems that some ZIP algorithms don't do this, creating file entries without directory entries.

Btw, the ZIP that causes this can be unzipped by typical tools like `unzip` without any problem so it seems to be a common occurrence.

### Error log

```
Caused by: java.nio.file.NoSuchFileException: /tmp/projectDir1890198260116576501/app/.gitignore
	at sun.nio.fs.UnixException.translateToIOException(UnixException.java:86)
	at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:102)
	at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:107)
	at sun.nio.fs.UnixFileSystemProvider.newByteChannel(UnixFileSystemProvider.java:214)
	at java.nio.file.spi.FileSystemProvider.newOutputStream(FileSystemProvider.java:434)
	at java.nio.file.Files.newOutputStream(Files.java:216)
	at java.nio.file.Files.copy(Files.java:3016)
	at io.fabric8.launcher.base.Paths.unzip(Paths.java:61)
	at io.fabric8.launcher.web.endpoints.LaunchEndpoint.uploadZip(LaunchEndpoint.java:107)
	at io.fabric8.launcher.web.endpoints.LaunchEndpoint$Proxy$_$$_WeldSubclass.uploadZip$$super(Unknown Source)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.jboss.weld.interceptor.proxy.TerminalAroundInvokeInvocationContext.proceedInternal(TerminalAroundInvokeInvocationContext.java:49)
	at org.jboss.weld.interceptor.proxy.AroundInvokeInvocationContext.proceed(AroundInvokeInvocationContext.java:77)
	at org.hibernate.validator.internal.cdi.interceptor.ValidationInterceptor.validateMethodInvocation(ValidationInterceptor.java:79)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.jboss.weld.interceptor.reader.SimpleInterceptorInvocation$SimpleMethodInvocation.invoke(SimpleInterceptorInvocation.java:73)
	at org.jboss.weld.interceptor.proxy.InterceptorMethodHandler.executeAroundInvoke(InterceptorMethodHandler.java:84)
	at org.jboss.weld.interceptor.proxy.InterceptorMethodHandler.executeInterception(InterceptorMethodHandler.java:72)
	at org.jboss.weld.interceptor.proxy.InterceptorMethodHandler.invoke(InterceptorMethodHandler.java:56)
	at org.jboss.weld.bean.proxy.CombinedInterceptorAndDecoratorStackMethodHandler.invoke(CombinedInterceptorAndDecoratorStackMethodHandler.java:79)
	at org.jboss.weld.bean.proxy.CombinedInterceptorAndDecoratorStackMethodHandler.invoke(CombinedInterceptorAndDecoratorStackMethodHandler.java:68)
	at io.fabric8.launcher.web.endpoints.LaunchEndpoint$Proxy$_$$_WeldSubclass.uploadZip(Unknown Source)
	at io.fabric8.launcher.web.endpoints.LaunchEndpoint$Proxy$_$$_WeldClientProxy.uploadZip(Unknown Source)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.jboss.resteasy.core.MethodInjectorImpl.invoke(MethodInjectorImpl.java:140)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invokeOnTarget(ResourceMethodInvoker.java:295)
	... 58 more

```

### PR DESCRIPTION
**Pull request by quintesse**
With ZIP files like that you can get `FileNotFoundException`
because the target directory doesn't exist yet.

Closes #543

- [x] I followed the contribution [guidelines](https://raw.githubusercontent.com/fabric8-launcher/launcher-backend/master/CONTRIBUTING.md).
- [x] I created an [issue](https://github.com/fabric8-launcher/launcher-backend/issues/new) and used it in the commit message.
- [x] I have built the project locally prior to submission with `mvn clean install`.
- [x] I kept a clean commit log (no unnecessary commits, no merge commits).
- [x] I am happy with my contribution.

### CODE REVIEW

**Comment by quintesse** [comment]
> Do you have a test case with a ZIP file illustrating this scenario? It seems weird that the entry wasn't previously created in line 59

@gastaldi it's because you can't assume that a ZIP will always have directory entries.

I thought adding a test would be hard because I was streaming the generated zip from one endpoint to the other, but actually the same error occurs when creating a file instead, so that's good :)

So I updated the PR with a test. The zip included in the test displays no problems whatsoever with any of the tools I tried to open it with, no warnings, nothing. So to those tools the file seems completely fine. Which to me seems to suggest that the problem was in our algorithm.

File: base/src/test/java/io/fabric8/launcher/base/PathsTest.java [hunk 1]
```diff
@@ -55,5 +55,15 @@ public void unzip() throws IOException {
...
     }
 
+    /*
+     * See https://github.com/fabric8-launcher/launcher-backend/pull/541
+     */
+    @Test
+    public void unzip_zip_without_dir_entries() throws IOException {
+        Path tempDirectory = temporaryFolder.newFolder().toPath();
+        Files.delete(tempDirectory);
+        Paths.unzip(new FileInputStream("src/test/resources/io/fabric8/launcher/base/pathtest.zip"), tempDirectory);
```
Comment (gastaldi) [inline base/src/test/java/io/fabric8/launcher/base/PathsTest.java]: `try (InputStream is = getClass().getResourceAsStream("pathtest.zip")){
...
}`

Comment (gastaldi) [reply base/src/test/java/io/fabric8/launcher/base/PathsTest.java]: `assertThatCode(()->Paths.unzip).doesNotThrowAnyException();

Comment (quintesse) [reply base/src/test/java/io/fabric8/launcher/base/PathsTest.java]: Ok done, but isn't the assert implicit in the test?

**Comment by quintesse** [comment]
The build is again failing due to external factors unrelated to the PR.

**Pull request merged**